### PR TITLE
Linear make prediction

### PR DESCRIPTION
**Pull request by Gordon119**
## What does this PR do?

as title.

## Test CLI & API (`bash tests/autotest.sh`)
- [x] Test Pass
  - (Copy and paste the last outputted line here.)
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Test quickstart & API (`bash tests/docs/test_examples.sh`)
- [x] Test Pass
  - (Copy and paste the last outputted line here.)
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Check API Document
- [x] API document is updated ([linear](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/linear.html), [nn](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/nn.html))
- [ ] Not Applicable (i.e., the PR does not include API changes.)

### CODE REVIEW

File: libmultilabel/linear/linear.py [hunk 1]
```diff
@@ -540,3 +540,20 @@ def predict_values(model, x: sparse.csr_matrix) -> np.ndarray:
...
+        x (np.ndarray): A matrix of decision values with dimension number of instances * number of classes.
+        top_k (int): Determine how many classes per instance should be predicted.
+
+    Returns:
+        list of tuples which contain predicted top k labels.
+    """
+    top_k_ind = np.argpartition(x, -top_k)[:, -top_k:]
+    pred_result = np.zeros(x.shape)
+    np.put_along_axis(pred_result, top_k_ind, 1, -1)
+    pred_result[pred_result != 1] = 0
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:558]: why line 588 as pred_result is initialized as the zero matrix?

File: libmultilabel/linear/linear.py [hunk 2]
```diff
@@ -540,3 +540,20 @@ def predict_values(model, x: sparse.csr_matrix) -> np.ndarray:
         ], 'csr')
 
     return (x * model['weights']).A + model['threshold']
+
+def predict_topk(preprocessor, x: np.ndarray, top_k: int = 5) -> 'list[tuple(str)]':
+    """Make top k predictions from decision values.
+
+    Args:
+        preprocessor: The preprocessor object from ``Preprocessor`` API used to load and preprocess the data.
+        x (np.ndarray): A matrix of decision values with dimension number of instances * number of classes.
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:549]: x should be called preds; see linear/metrics.py

File: libmultilabel/linear/linear.py (hunk 1 above)
Comment (Gordon119) [reply libmultilabel/linear/linear.py]: ``inverse_transform`` has to take predictions in binary labels, so I first set pred_result to zero matrix and set the top k labels to 1. I realized that line 558 is redundant and removed it.

File: libmultilabel/linear/linear.py [hunk 3]
```diff
@@ -540,3 +540,19 @@ def predict_values(model, x: sparse.csr_matrix) -> np.ndarray:
...
+        preds (np.ndarray): A matrix of decision values with dimension number of instances * number of classes.
+        top_k (int): Determine how many classes per instance should be predicted.
+
+    Returns:
+        list of tuples which contain predicted top k labels.
+    """
+    top_k_ind = np.argpartition(preds, -top_k)[:, -top_k:]
+    pred_result = np.zeros(preds.shape)
+    np.put_along_axis(pred_result, top_k_ind, 1, -1)
+    return preprocessor.binarizer.inverse_transform(pred_result)
```
Comment (cjlin1) [inline libmultilabel/linear/linear.py:558]: where do we have this binarizer.inverse_transform defined??

Comment (Gordon119) [reply libmultilabel/linear/linear.py:558]: Since we’re using sklearn’s MultiLabelBinarizer, I used this:
https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MultiLabelBinarizer.html

File: libmultilabel/linear/linear.py [hunk 4]
```diff
@@ -540,3 +540,19 @@ def predict_values(model, x: sparse.csr_matrix) -> np.ndarray:
         ], 'csr')
 
     return (x * model['weights']).A + model['threshold']
+
```
Comment (Eleven1Liu) [inline libmultilabel/linear/linear.py:543]:
```suggestion


```

**Comment by Eleven1Liu** [comment]
@Sinacam review

File: libmultilabel/linear/linear.py [hunk 5]
```diff
@@ -4,6 +4,7 @@
 import numpy as np
 import scipy.sparse as sparse
 from liblinear.liblinearutil import train
+from .preprocessor import Preprocessor
```
Comment (Eleven1Liu) [inline libmultilabel/linear/linear.py:7]: Please sort imports and remove unused ones (Preprocessor).

**Review by Eleven1Liu — CHANGES_REQUESTED** [review]

File: libmultilabel/linear/linear.py [hunk 6]
```diff
@@ -540,3 +540,20 @@ def predict_values(model, x: sparse.csr_matrix) -> np.ndarray:
...
+
+    Args:
+        label_mapping (np.ndarray): A ndarray of class labels that maps each label to its index. 
+        preds (np.ndarray): A matrix of decision values with dimension number of instances * number of classes.
+        top_k (int): Determine how many classes per instance should be predicted.
+
+    Returns:
+        list of lists which contain top k labels.
+    """
+    top_k_ind = np.argpartition(preds, -top_k)[:, -top_k:]
```
Comment (Sinacam) [inline libmultilabel/linear/linear.py:556]: `top_k_ind` is sorted ascending this way. `top_k_ind = np.argpartition(preds, -top_k)[:, :top_k:-1]` sorts descending.

File: libmultilabel/linear/linear.py [hunk 7]
```diff
@@ -540,3 +540,20 @@ def predict_values(model, x: sparse.csr_matrix) -> np.ndarray:
...
+    pred_result = np.zeros(preds.shape)
+    np.put_along_axis(pred_result, top_k_ind, 1, -1)
+    return [list(label_mapping.compress(row)) for row in pred_result]
```
Comment (Sinacam) [inline libmultilabel/linear/linear.py:559]: Isn't 557-559 the same as `return label_mapping[top_k_ind].tolist()`?

**Comment by Eleven1Liu** [comment]
LGTM

**Review by Sinacam — APPROVED** [review]

**Review by Eleven1Liu — APPROVED** [review]
